File: cli/golem-cli/templates/scala/streaming/component-dir/src/main/scala/component_name/StreamingAgent.scala
package component_name

import golem.BaseAgent
import golem.config.AgentConfig
import golem.runtime.annotations.{agentDefinition, durableStreamSlot, durableStreams, endpoint}
import zio.blocks.schema.Schema
import golem.config.Secret

import scala.concurrent.Future

final case class StreamingConfig(externalAuth: Secret[String]) derives Schema

@agentDefinition(mount = "/durable-stream-agents/{name}")
trait StreamingAgent extends BaseAgent with AgentConfig[StreamingConfig] {
  class Id(val name: String)

  def sum(input: golem.schema.AgentStream[Int]): Future[Int]
  def produce(): Future[golem.schema.AgentStream[Int]]
  def transform(
    prefix: String,
    input: golem.schema.AgentStream[Int]
  ): Future[golem.schema.AgentStream[String]]
  def nested(): Future[golem.schema.AgentStream[golem.schema.AgentStream[Int]]]
  def recoverable(): Future[golem.schema.AgentStream[Either[String, Int]]]
  def status(): Future[String]

  @endpoint(method = "PUT", path = "/echo")
  @durableStreamSlot(source = "input", slot = "input")
  @durableStreamSlot(source = "output", slot = "$result")
  @durableStreams(allowExternalWrites = true)
  def durableEcho(input: golem.schema.AgentStream[String]): Future[golem.schema.AgentStream[String]]

  def appendExternal(
    url: String,
    producerId: String,
    values: Vector[String],
    close: Boolean
  ): Future[Option[String]]

  def readExternal(url: String): Future[Vector[String]]
}

File: cli/golem-cli/templates/effect/streaming/component-dir/src/streaming-agent.ts
import { Effect, Option, Ref, Result, Schema, Stream } from "effect";
import {
  defineAgent,
  defineConfig,
  DurableStreams,
  Http,
  method,
  WitTypes,
} from "@golemcloud/effect-golem";

class StreamingConfig extends defineConfig("StreamingAgent.Config", {
  externalAuth: Schema.Redacted(Schema.String),
}) {}

export const StreamingAgent = defineAgent({
  name: "StreamingAgent",
  id: { name: Schema.String },
  config: StreamingConfig,
  http: Http.mount("/durable-stream-agents/{name}"),
  methods: {
    sum: method({
      input: { input: WitTypes.AgentStream(Schema.Number) },
      success: Schema.Number,
    }),
    produce: method({ input: {}, success: WitTypes.AgentStream(Schema.Number) }),
    transform: method({
      input: {
        prefix: Schema.String,
        input: WitTypes.AgentStream(Schema.Number),
      },
      success: WitTypes.AgentStream(Schema.String),
    }),
    nested: method({
      input: {},
      success: WitTypes.AgentStream(WitTypes.AgentStream(Schema.Number)),
    }),
    recoverable: method({
      input: {},
      success: WitTypes.AgentStream(Schema.Result(Schema.Number, Schema.String)),
    }),
    status: method({ input: {}, success: Schema.String }),
    durableEcho: method({
      input: { input: WitTypes.AgentStream(Schema.String) },
      success: WitTypes.AgentStream(Schema.String),
      http: [
        Http.put("/echo", {
          durableStreams: {
            slots: [
              { source: "input", slot: "input" },
              { source: "output", slot: "$result" },
            ],
            allowExternalWrites: true,
          },
        }),
      ],
    }),
    appendExternal: method({
      input: {
        url: Schema.String,
        producerId: Schema.String,
        values: Schema.Array(Schema.String),
        close: Schema.Boolean,
      },
      success: Schema.Option(Schema.String),
    }),
    readExternal: method({
      input: { url: Schema.String },
      success: Schema.Array(Schema.String),
    }),
  },
}).implement({
  init: () => Ref.make(0),
  methods: (cancelledProducers) => ({
    sum: ({ input }) =>
      input.pipe(
        Stream.runFold(
          () => 0,
          (total, value) => total + value,
        ),
        Effect.orDie,
      ),
    produce: () => Effect.succeed(Stream.make(1, 2, 3)),
    transform: ({ prefix, input }) =>
      Effect.succeed(input.pipe(Stream.map((value) => `${prefix}:${value}`))),
    nested: () =>
      Effect.succeed(Stream.make(Stream.make(10, 20), Stream.make(30, 40))),
    recoverable: () =>
      Effect.succeed(
        Stream.make(
          Result.succeed(1),
          Result.fail("this item could not be produced"),
          Result.succeed(2),
        ),
      ),
    status: () =>
      Ref.get(cancelledProducers).pipe(
        Effect.map((count) => `ready (${count} cancelled producers)`),
      ),
    durableEcho: ({ input }) =>
      Effect.succeed(input.pipe(Stream.map((value) => `echo:${value}`))),
    appendExternal: ({ url, producerId, values, close }) =>
      Effect.gen(function* () {
        const config = yield* StreamingConfig;
        const auth = yield* config.externalAuth.borrow;
        const writer = yield* DurableStreams.makeJsonWriter(Schema.String, {
          url,
          producerId,
          auth,
        });
        const receipt = yield* writer.append(values, { close });
        return Option.fromUndefinedOr(receipt.nextOffset);
      }).pipe(Effect.orDie, Effect.scoped),
    readExternal: ({ url }) =>
      Effect.gen(function* () {
        const config = yield* StreamingConfig;
        const auth = yield* config.externalAuth.borrow;
        const values = yield* DurableStreams.readJson(Schema.String, {
          url,
          auth,
        }).pipe(Stream.runCollect);
        return Array.from(values);
      }).pipe(Effect.orDie, Effect.scoped),
  }),
});
